enterprise/server/cmd/action_runner/main.go

```diff
@@ -0,0 +1,662 @@
+		Id: &bespb.BuildEventId{Id: &bespb.BuildEventId_Started{Started: &bespb.BuildEventId_BuildStartedId{}}},
+		Children: []*bespb.BuildEventId{
+			{Id: &bespb.BuildEventId_Progress{Progress: &bespb.BuildEventId_ProgressId{OpaqueCount: 0}}},
+			{Id: &bespb.BuildEventId_WorkspaceStatus{WorkspaceStatus: &bespb.BuildEventId_WorkspaceStatusId{}}},
+			{Id: &bespb.BuildEventId_BuildFinished{BuildFinished: &bespb.BuildEventId_BuildFinishedId{}}},
+		},
+		Payload: &bespb.BuildEvent_Started{Started: &bespb.BuildStarted{
+			Uuid:            ar.bep.streamID.InvocationId,
+			StartTimeMillis: time.Now().UnixNano() / int64(time.Millisecond),
+		}},
+	}
+	if err := bep.Publish(startedEvent); err != nil {
+		return nil
+	}
+	if err := ar.flushProgress(); err != nil {
+		return nil
+	}
+	workspaceStatusEvent := &bespb.BuildEvent{
+		Id: &bespb.BuildEventId{Id: &bespb.BuildEventId_WorkspaceStatus{WorkspaceStatus: &bespb.BuildEventId_WorkspaceStatusId{}}},
+		Payload: &bespb.BuildEvent_WorkspaceStatus{WorkspaceStatus: &bespb.WorkspaceStatus{
+			Item: []*bespb.WorkspaceStatus_Item{
+				{Key: "BUILD_USER", Value: ar.username},
+				{Key: "BUILD_HOST", Value: ar.hostname},
+				{Key: "REPO_URL", Value: *repoURL},
+				{Key: "COMMIT_SHA", Value: *commitSHA},
+				{Key: "GIT_TREE_STATUS", Value: "Clean"},
+				// TODO: Populate GIT_BRANCH. Can't source this from the `trigger_branch` flag
+				// in the PR case, because that refers to the branch into which the PR would be
+				// merged, which doesn't reflect the currently checked out branch.
+			},
+		}},
+	}
+	if err := bep.Publish(workspaceStatusEvent); err != nil {
+		return nil
+	}
+
+	ps1End := "$"
+	if ar.username == "root" {
+		ps1End = "#"
+	}
+
+	stopFlushingProgress := ar.startBackgroundProgressFlush()
+	defer stopFlushingProgress()
+
+	for _, bazelCmd := range ar.action.BazelCommands {
+		args, err := bazelArgs(bazelCmd)
+		// Provide some info to help make it clear which output is coming
+		// from which bazel commands.
+		ar.log.Printf("\n%s%s@%s%s%s bazelisk %s", textCyan, ar.username, ar.hostname, textReset, ps1End, strings.Join(args, " "))
+		if err != nil {
+			return fmt.Errorf("failed to parse bazel command: %s", err)
+		}
+		err = runCommand(ctx, "bazelisk", args, emptyEnv, ar.log)
+		if exitCode := getExitCode(err); exitCode != noExitCode {
+			ar.log.Printf("%s(command exited with code %d)%s", textGray, exitCode, textReset)
+		}
+		// If the command failed, report its progress before returning.
+		if err := ar.flushProgress(); err != nil {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ar *actionRunner) startBackgroundProgressFlush() func() {
+	stop := make(chan struct{}, 1)
+	go func() {
+		for {
+			select {
+			case <-stop:
+				break
+			case <-time.After(progressFlushInterval):
+				ar.flushProgress()
+			}
+		}
+	}()
+	return func() {
+		stop <- struct{}{}
+	}
+}
+
+func (ar *actionRunner) flushProgress() error {
```

Comment:
A nicer way to do this is to flush the buffer after it accumulates a certain number of bytes. Then, if you still find you need more interactivity, you could layer on a timer that says also publish every second or something.

The risk of running it on a timer only is that you might publish thousands (or more?!) of very small buffers.